org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/PasteEventHandler.java

```diff
@@ -334,6 +343,77 @@ private static String getEol(String text) {
+		String desiredPath = "";
+		final ASTParser parser = ASTParser.newParser(AST.getJLSLatest());
+		parser.setResolveBindings(false);
+		parser.setKind(ASTParser.K_COMPILATION_UNIT);
+		parser.setBindingsRecovery(false);
+		parser.setIgnoreMethodBodies(true);
+		parser.setEnvironment(new String[0], new String[0], null, true);
+		Map<String, String> javaOptions = JavaCore.getOptions();
+		parser.setCompilerOptions(javaOptions);
+		parser.setSource(content.toCharArray());
+		CompilationUnit root = (CompilationUnit) parser.createAST(monitor);
+		// Check for package, import, and/or type declaration to determine if java file
+		if (root != null && (root.getPackage() != null || !root.imports().isEmpty() || !root.types().isEmpty())) {
+			String pathRoot = path;
+			if (root.getPackage() != null) {
+				String packageName = root.getPackage().getName().toString();
+				IJavaProject[] projects = ProjectUtils.getJavaProjects();
+				boolean pathFound = false;
+				String partialPathAddition = "";
+				for (int i = 0; i < projects.length; i++) {
+					if (!ProjectsManager.DEFAULT_PROJECT_NAME.equals(projects[i].getElementName()) && !pathFound) {
+						IPackageFragment[] packages = projects[i].getPackageFragments();
+						for (int j = 0; j < packages.length; j++) {
+							if (packages[j].getKind() == IPackageFragmentRoot.K_SOURCE && !packages[j].isDefaultPackage()) {
+								pathRoot = packages[j].getParent().getResource().getLocation().toOSString();
+								String name = packages[j].getElementName();
+								if (name.equals(packageName)) {
+									desiredPath = packages[j].getResource().getLocation().toOSString();
+									pathFound = true;
+									break;
+								} else if (packageName.startsWith(name) && packageName.substring(0, name.length() + 1).endsWith(".")) {
+									String tempPath = packages[j].getResource().getLocation().toOSString();
+									if (tempPath.length() > desiredPath.length()) {
+										desiredPath = tempPath;
+										partialPathAddition = packageName.substring(name.length()).replaceAll("\\.", File.separator);
+									}
+								}
+							}
+						}
+					}
+				}
+
+				if (!pathFound && !partialPathAddition.isEmpty()) {
+					desiredPath = desiredPath + partialPathAddition;
+				} else if (!pathFound) {
+					packageName = packageName.replaceAll("\\.", File.separator);
+					desiredPath = pathRoot + File.separator + packageName;
+				}
+			} else {
+				desiredPath = pathRoot;
+			}
+
+			String possibleFileName = "Untitled";
+			if (!root.types().isEmpty()) {
+				possibleFileName = ((TypeDeclaration) root.types().get(0)).getName().toString();
+			}
+			int counter = 1;
+			File testExists = new File(desiredPath + File.separator + possibleFileName + ".java");
+			String finalFileName = possibleFileName;
+			while (testExists.exists()) {
+				finalFileName = possibleFileName + Integer.toString(counter);
+				testExists = new File(desiredPath + File.separator + finalFileName + ".java");
+				counter++;
+			}
```

Comment:
This can be written a bit more simply. Use something like `Files.exists(Paths.get(desiredPath, possibleFileName, Integer.toString(counter)))` . When it returns `true`, you can just return that.

Let's just say it's the client's responsibility to verify that the returned path is writable since the server might not know what the client can/can't do (they often are 2 separate processes).